File: CS2370/groceries/groceries/groceries.cpp
// groceries_template.cpp: Stores Orders in a list.

#include <fstream>
#include <iostream>
#include <list>
#include <stdexcept>
#include <sstream>
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>
#include <iomanip>
//#include "split.h"
using namespace std;

//////////////////
// Customer code /
//////////////////
struct Customer {
    int cust_id;
    string name;
    string street;
    string city;
    string state;
    string zip;
    string phone;
    string email;
    Customer(int id, const string& name, const string& street, const string& city,
        const string& state, const string& zip, const string& phone, const string& email)
        : name(name), street(street), city(city), state(state), zip(zip), phone(phone), email(email)
    {
        cust_id = id;
    }
    string print_detail() const {
        string string_detail;
        string_detail = (
            "Customer ID #" + to_string(cust_id) + ":" +"\n"
            + name + ", ph. " + phone + ", " + "email: " + email + "\n"
            + street + "\n" + city + ", " + state + " " + zip + "\n"
            );
        return string_detail;
    }
};

vector<Customer> customers;
void read_customers(const string& fname) {
    ifstream customer_data;
    stringstream line_of_data;
    string piece_of_data;
    vector<string> customer;
    customer_data.open(fname);
    if (!customer_data.is_open()) {
        cout << "Failed to open file" << endl;
        return;
    }
    while (!customer_data.fail()) {
        getline(customer_data, piece_of_data);
        line_of_data.clear();
        line_of_data.str(piece_of_data);
        if (piece_of_data.empty()) {
        }
        else {
            while (!line_of_data.eof()) {
                getline(line_of_data, piece_of_data, ',');
                customer.push_back(piece_of_data);
            }
            customers.push_back(Customer(stoi(customer.at(0)),
                customer.at(1), customer.at(2), customer.at(3),
                customer.at(4), customer.at(5), customer.at(6), customer.at(7)));
            customer.clear();
        }
    }
    if (!customer_data.eof()) {
        cout << "Failed to read all of the data in the file" << endl;
        return;
    }
    customer_data.close();
}

int find_cust_idx(int cust_id) {
    for (int i = 0; i < customers.size(); ++i)
        if (cust_id == customers[i].cust_id)
            return i;
    throw runtime_error("Customer not found");
}

//////////////
// Item code /
//////////////
struct Item {
    int item_id;
    string description;
    double price;
    Item(int id, const string& desc, double pric) : description(desc) {
        item_id = id;
        price = pric;
    }

};

vector<Item> items;
void read_items(const string& fname) {
    ifstream item_data;
    stringstream line_of_data;
    string piece_of_data;
    vector<string> item_lyst;
    item_data.open(fname);
    if (!item_data.is_open()) {
        cout << "Failed to open file" << endl;
        return;
    }
    while (!item_data.fail()) {
        getline(item_data, piece_of_data);
        line_of_data.clear();
        line_of_data.str(piece_of_data);
        if (piece_of_data.empty()) {
        }
        else {
            while (!line_of_data.eof()) {
                getline(line_of_data, piece_of_data, ',');
                item_lyst.push_back(piece_of_data);
            }
            items.push_back(Item(stoi(item_lyst.at(0)), item_lyst.at(1), stod(item_lyst.at(2))));
            item_lyst.clear();
            item_data.clear();
        }
    }
    if (!item_data.eof()) {
        cout << "Failed to read all of the data in the file" << endl;
        return;
    }
    item_data.close();
   
}

int find_item_idx(int item_id) {
    for (int i = 0; i < items.size(); ++i)
        if (item_id == items[i].item_id)
            return i;
    throw runtime_error("Item not found");
}

class LineItem {
    int item_id;
    int qty;
    friend class Order;
public:
    LineItem(int id, int q) {
        item_id = id;
        qty = q;
    }
    double sub_total() const {
        int idx = find_item_idx(item_id);
        return items[idx].price * qty;
    }
    friend bool operator<(const LineItem& item1, const LineItem& item2) {
        return item1.item_id < item2.item_id;
    }
};

/////////////////
// Payment code /
/////////////////
class Payment {
    double amount = 0.0;
    friend class Order;
public:
    Payment() = default;
    virtual ~Payment() = default;
    virtual string print_detail() const = 0;
};

class Credit : public Payment {
    string card_number;
    string expiration;
public:
    Credit(string card_number_input, string expiration_input) {
        card_number = card_number_input;
        expiration = expiration_input;
        
    }
    string print_detail() const {
        string payment_info;
        payment_info = "Paid by Credit card " + card_number + ", exp. " + expiration;
        return payment_info;

    }
};

class Paypal : public Payment {
    string paypal_id;
public:
    Paypal(const string paypal_id_input) {
        paypal_id = paypal_id_input;
    }

    string print_detail() const {
        string payment_info;
        payment_info = "Paid by Paypal ID: " + paypal_id;
        return payment_info;

    }
};

class WireTransfer : public Payment {
    string bank_id;
    string account_id;
public:
    WireTransfer(const string bank_id_input, const string account_id_input) {
        bank_id = bank_id_input;
        account_id = account_id_input;
    }

    string print_detail() const {
        string payment_info;
        payment_info = "Paid by Wire transfer from Bank ID " + bank_id + ", Account# " + account_id;
        return payment_info;

    }
};

///////////////
// Order code /
///////////////
class Order {
    int order_id;
    string order_date;
    int cust_id;
    vector<LineItem> line_items;
    Payment* payment;
    double order_total = 0.0;
public:
    Order(int c_id, int id, const string& date, const vector<LineItem>& items,Payment* p)
        : order_date(date), line_items(items) {
        order_id = id;
        cust_id = c_id;
        payment = p;
        sort(line_items.begin(), line_items.end());
        for (auto i : line_items) {
            order_total += i.sub_total();
        }

    }
    ~Order() {
        delete payment;
    }
    double total() const {
        return order_total;
    }
    string print_detail() const {
        string string_detail;
        string order_string;

        order_string = to_string(order_total);
        order_string.erase(order_string.find_last_not_of('0') + 1, string::npos);
        string_detail = ("=========================== \nOrder # "
            + to_string(order_id) + ", Date: " + to_string(this->total()) + "\n"
            + "Amount: " + order_string + ", " + payment->print_detail() 
            + "\n" + customers[find_cust_idx(cust_id)].print_detail() + "\n"
            + "Order Detail: \n");
        for (auto i : line_items) {
            string str_price;
            str_price = to_string(items[find_item_idx(i.item_id)].price);
            str_price.erase(str_price.find_last_not_of('0') + 1, string::npos);
            string_detail += "\t Item " 
                + to_string(i.item_id) + ": \"" 
                + items[find_item_idx(i.item_id)].description 
                + "\", " + to_string(i.qty) 
                + " @ " + str_price + "\n";
        }
        return string_detail;
    }
};

list<Order> orders;

void read_orders(const string& fname) {
    ifstream orderf(fname);
    string line;
    string line2;
    vector<string> temp_order_lyst;
    vector<string> order_item;
    vector<string> payment_items;
    vector<LineItem> line_items;
    vector<Payment*> payment_lyst;
    stringstream line_of_data;
    stringstream item_data;
    stringstream payment_line;
    string item_temp;
    int order_id;
    string order_date;
    int cust_id;
    bool test;
    test = true;
    Payment* pmt;
    while (getline(orderf, line)) {
        line_of_data.str(line);
        line_of_data.clear();
        while (!line_of_data.eof()) {
            getline(line_of_data, line2, ',');
            temp_order_lyst.push_back(line2);
        }
        if (temp_order_lyst.empty() || test == false) {
            test = true;
            payment_line.str(line);
            payment_line.clear();
            while (!payment_line.eof()) {
                getline(payment_line, line, ',');
                payment_items.push_back(line);
            }
            if (stoi(payment_items[0]) == 1) {
                
                payment_lyst.push_back(new Credit(payment_items[1], payment_items[2]));
            }
            else if (stoi(payment_items[0]) == 2) {
                payment_lyst.push_back(new Paypal(payment_items[1]));

            }
            else if (stoi(payment_items[0]) == 3) {
                payment_lyst.push_back(new WireTransfer(payment_items[1], payment_items[2]));
            }
            else {
                cout << "Error";
            }
            payment_items.clear();
            temp_order_lyst.clear();
        }
        else {
            cust_id = stoi(temp_order_lyst[0]);
            order_id = stoi(temp_order_lyst[1]);
            order_date = temp_order_lyst[2];
            for (int i = 3; i < temp_order_lyst.size(); i++) {
                item_data.str(temp_order_lyst[i]);
                item_data.clear();
                while (!item_data.eof()) {
                    getline(item_data, item_temp, '-');
                    order_item.push_back(item_temp);
                }
                line_items.push_back(LineItem(stoi(order_item[0]), stoi(order_item[1])));
                order_item.clear();

            }
            temp_order_lyst.clear();
            test = false;
        }

        if (payment_lyst.empty()) {

        }
        else {
            pmt = payment_lyst[0];

            orders.emplace_back(cust_id, order_id, order_date, line_items, pmt);
            line_items.clear();
            payment_lyst.erase(payment_lyst.begin());


        }


    }
}

int main() {
    read_customers("customers.txt");
    read_items("items.txt");
    read_orders("orders.txt");

    ofstream ofs("order_report.txt");
    for (const Order& order : orders)
        ofs << order.print_detail() << endl;
}

File: CS3270/Project-D/groceries.py
"""
    Cody Strange
    11/10/2022
    CS3270-001
    Programming project D

    Summary:
            This program will print a report of online orders
            from a fictional grocery chain using data from customers.txt, orders.txt, and items.txt

            Report:
                - <Order #>, <date>
                - <Amount>, <method of >
                - <Customer ID>
                - <Name>, <phone-number>, <email>
                - <Address>
                - <City>, <State> <zip-code>
                - Order Detail:
                    <Item #>: <"item">, <quantity> @ <price>

    Bugs:
        - none found

    Required features:
        - none remaining

    Recomended features
        - create more test cases for remaining code
        - create multiple files to hold the code
"""

class Customer():
    """Customer Object
            fields
                cust_id:string
                name:string
                street:string
                city:string
                state:string
                postal_code:string
                phone:string
                email:string
            Methods:
                __init__(...)
                __str__():string
                read_customers(fname:string) 
    """
    def __init__(self, cust_id, name, street, city, state, postal_code, phone, email):
        self.cust_id = cust_id
        self.name = name
        self.street = street
        self.city = city
        self.state = state
        self.postal_code = postal_code
        self.phone = phone
        self.email = email

    def __str__(self):
        customer_string = f"{self.cust_id},{self.name},{self.street},{self.city},{self.state},{self.postal_code},{self.phone},{self.email}"
        return customer_string
    
    @staticmethod
    def read_customer(fname):
        with open(fname, 'r') as file:
            return file.readlines()

class Order():
    """Order Object
            fields
                order_id:string
                order_date:string
                cust_id:string
                line_items:list
                payment:Payment
            Methods:
                __str__():string
                total:float(@property)
                read_orders(fname:string)
    """
    def __init__(self, cust_id, order_id, order_date, line_items, payment):
        self.cust_id = cust_id
        self.order_id = order_id
        self.order_date = order_date
        self.line_items = line_items
        self.payment = payment
    @property
    def total():
        pass
    @staticmethod
    def read_orders(fname):
        with open(fname, 'r') as file:
            return file.readlines()
    def __str__(self):
        line_items_string = ""
        for i in self.line_items : line_items_string += f"{i},"
        order_string = f"{self.cust_id},{self.order_id},{self.order_date},{line_items_string}{self.payment}"
        return order_string

class PayPal():
    """PayPal Object
            fields
                paypal_id:string
            methods
                __str__():string
    """
    def __init__(self, paypal_id):
        super().__init__()
        self.paypal_id = paypal_id

    def __str__(self):
        paypal_string = f"{self.paypal_id}"
        return paypal_string

class Wire():
    """Wire Object
        fields
            bank_id:string
            account_id:string
        methods
            __str__():string
    """
    def __init__(self, bank_id, account_id):
        super().__init__()
        self.bank_id = bank_id
        self.account_id = account_id
    def __str__(self):
        wire_string = f"{self.bank_id},{self.account_id}"
        return wire_string

class Credit():
    """Credit Object
            fields
                card_number:string
                exp_date:string
            methods
                __str__():string
    """
    def __init__(self, card_number, exp_date):
        super().__init__()
        self.card_number = card_number
        self.exp_date = exp_date
    def __str__(self):
        credit_string = f"{self.card_number},{self.exp_date}"
        return credit_string

class LineItem():
    """LineItem Object
            fields
                item_id:string
                qty:int
            methods
                sub_total():float
    """
    def __init__(self, item_id, qty):
        self.item_id = item_id
        self.qty = qty
    def sub_total(self):
        line_item_string = f"{self.item_id},{self.qty}"
        return line_item_string

class Item():
    """Item Object
        fields
            item_id:string
            description:string
            price:float
        methods
            read_items(fname:string)
    """
    def __init__(self, item_id, description, price):
        self.item_id = item_id
        self.description = description
        self.price = price
    @staticmethod
    def read_items(fname):
        with open(fname, 'r') as file:
            return file.readlines()
    def __str__(self):
        item_string = f"{self.item_id},{self.description},{self.price}"
        return item_string

def line_items(order_data):
    return order_data[3:]

def customer_database(customer_file):
    customers = {}
    for customer in customer_file:
        customer_data = customer.strip().split(',')
        cid = customer_data[0]
        cname = customer_data[1]
        cstreet = customer_data[2]
        ccity = customer_data[3]
        cstate = customer_data[4]
        cpostal_code = customer_data[5]
        cphone = customer_data[6]
        cemail = customer_data[7]
        customers[cid] = Customer(cid, cname, cstreet, ccity, cstate, cpostal_code, cphone, cemail)
    return customers

def order_database(orders_file):
    order = {}
    order_lyst = [x for x in orders_file if orders_file.index(x) % 2 == 0]
    opayment_lyst = [x for x in orders_file if orders_file.index(x) % 2 != 0]
    for opayment_line, order_line in zip(opayment_lyst, order_lyst):
        opayment_file = opayment_line.strip().split(',')
        opay_type = opayment_file[0]
        if opay_type == '1':
            ocredit_num = opayment_file[1]
            ocredit_expir = opayment_file[2]
            opayment = Credit(ocredit_num, ocredit_expir)
        elif opay_type == '2':
            opay_id = opayment_file[1]
            opayment = PayPal(opay_id)
        else:
            opay_bank = opayment_file[1]
            opay_account = opayment_file[2]
            opayment = Wire(opay_bank, opay_account)
        order_data = order_line.strip().split(',')
        ocid = order_data[0]       
        oid = order_data[1]
        odate = order_data[2]
        oline_items = line_items(order_data)
        order[oid] = Order(ocid, oid, odate, oline_items, opayment)
    return order

def items_database(items_file):
    items = {}
    for line in items_file:
        items_data = line.strip().split(',')
        iid = items_data[0]
        idescription = items_data[1]
        iprice = items_data[2]
        items[iid] = Item(iid, idescription, iprice)
    return items

def calc_total_amount(line_items, items):
    total = 0
    item_list = {}
    for item in line_items:
        item = item.split('-')
        item_list[item[0]] = LineItem(item[0], item[1])
    for id in item_list:
        item_price = float(items[id].price)
        item_qty = int(item_list[id].qty)
        total += (item_qty * item_price)
    return "{:.2f}".format(total)

def find_payment_info(payment):
    if isinstance(payment,PayPal):
        return f"Paid by PayPal ID: {payment.paypal_id}"
    elif isinstance(payment,Credit):
        return f"Paid by Credit card {payment.card_number}, exp. {payment.exp_date}"
    elif isinstance(payment,Wire):
        return f"Paid by Wire transfer from Bank ID {payment.bank_id}, Account# {payment.account_id}"
    else:
        return "FAILED"

def order_detail_page(line_items, items):
    order_detail_list = []
    item_list = {}
    for item in line_items:
        item = item.split('-')
        item_list[item[0]] = LineItem(item[0], item[1])
    for id in item_list:
        item_price = float(items[id].price)
        item_qty = int(item_list[id].qty)
        item_description = items[id].description
        order_detail_list.append(f'Item {id}: "{item_description}", {item_qty}, @ {item_price:.2f}')
    return order_detail_list

def main():
    orders = order_database(orders_file)
    items = items_database(items_file)
    customers = customer_database(customer_file)
    for order in orders:
        order_id = orders[order].order_id
        order_date = orders[order].order_date
        cust_id = orders[order].cust_id
        total_amount = calc_total_amount(orders[order].line_items, items)
        payment_information = find_payment_info(orders[order].payment)
        cust_name = customers[cust_id].name
        cust_phone = customers[cust_id].phone
        cust_email = customers[cust_id].email
        cust_street = customers[cust_id].street
        cust_city = customers[cust_id].city
        cust_state = customers[cust_id].state
        cust_postal_code = customers[cust_id].postal_code
        order_item_list = order_detail_page(orders[order].line_items, items)
        print(f"""==========================
Order#{order_id}, Date: {order_date}
Amount: ${total_amount}, {payment_information}

Customer ID #{cust_id}:
{cust_name}, ph. {cust_phone}, email: {cust_email}
{cust_street}
{cust_city}, {cust_state} {cust_postal_code}

Order Detail:""")

        for order_detail in order_item_list:
            print(f"\t{order_detail}")



if __name__ == '__main__':
 
    customer_file = Customer.read_customer("customers.txt")
    items_file = Item.read_items("items.txt")
    orders_file = Order.read_orders("orders.txt")
    main()


